pkg/dyninst/decode/decoder.go

```diff
@@ -0,0 +1,407 @@
+	)
+
+	frames, err = event.StackPCs()
+	if err != nil {
+		return err
+	}
+	enc := jsontext.NewEncoder(out)
+	err = enc.WriteToken(jsontext.BeginObject)
+	if err != nil {
+		return err
+	}
+
+	err = enc.WriteToken(jsontext.String("stack_frames"))
+	if err != nil {
+		return err
+	}
+	err = enc.WriteToken(jsontext.BeginArray)
+	if err != nil {
+		return err
+	}
+	for i := range frames {
+		err = enc.WriteToken(jsontext.Uint(frames[i]))
+		if err != nil {
+			return err
+		}
+	}
+	err = enc.WriteToken(jsontext.EndArray)
+	if err != nil {
+		return err
+	}
+
+	err = enc.WriteToken(jsontext.String("captures"))
+	if err != nil {
+		return err
+	}
+	err = enc.WriteToken(jsontext.BeginObject)
+	if err != nil {
+		return err
+	}
+	itemsByAddress := map[uint64]*dataItemWithAddrCounter{}
+	for item, err := range event.DataItems() {
+		if errors.Is(err, output.ErrFinishedIterating) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if eventRoot == nil {
+			// item is EventRoot
+			rootData = item.Data()
+			eventRoot, ok = d.program.Types[ir.TypeID(item.Header().Type)].(*ir.EventRootType)
+			if !ok {
+				return errors.New("expected event of type root first")
+			}
+			continue
+		}
+
+		itemsByAddress[item.Header().Address] = &dataItemWithAddrCounter{
+			dataItem:    item,
+			addrCounter: 0,
+		}
+	}
+
+	// We iterate over the 'Expressions' of the EventRoot which contains
+	// metadata and raw bytes of the parameters of this function.
+	for _, expr := range eventRoot.Expressions {
+		parameterType := expr.Expression.Type
+		parameterData := rootData[expr.Offset : expr.Offset+parameterType.GetByteSize()]
+		err = enc.WriteToken(jsontext.String(expr.Name))
+		if err != nil {
+			return err
+		}
+		err = d.encodeValue(enc, itemsByAddress, parameterType, parameterData)
+		if err != nil {
+			return fmt.Errorf("error parsing data for field %s: %w", eventRoot.Name, err)
+		}
+	}
+
+	err = enc.WriteToken(jsontext.EndObject) // captures
+	if err != nil {
+		return err
+	}
+	err = enc.WriteToken(jsontext.EndObject) // full object
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *Decoder) encodeValue(enc *jsontext.Encoder, itemsByAddress map[uint64]*dataItemWithAddrCounter, irType ir.Type, data []byte) error {
+	switch v := irType.(type) {
+	case *ir.BaseType:
+		return d.encodeBaseTypeValue(enc, v, data)
+	case *ir.PointerType:
+		if len(data) < int(v.GetByteSize()) {
+			return errors.New("passed data not long enough for pointer")
+		}
+		// handle nil pointer
+		addr := binary.NativeEndian.Uint64(data)
+		if addr == 0 {
+			err := enc.WriteToken(jsontext.String("nil"))
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+		pointedValue, ok := itemsByAddress[addr]
+		if !ok {
+			return errors.New("pointer not found in address pass map")
+		}
+		pointedValue.addrCounter++
+		if pointedValue.addrCounter > 1 {
+			return enc.WriteToken(jsontext.String(fmt.Sprintf("0x%x", pointedValue.dataItem.Header().Address)))
```

Comment:
nit: pull header and data out into variables -- if only because it'll make these very long lines shorter.